src/compiler/crystal/tools/typer.cr

```diff
@@ -0,0 +1,612 @@
+module Crystal
+  class SourceTyper
+    # Represents a fully typed definition signature
+    record Signature,
+      name : String,
+      return_type : Crystal::ASTNode,
+      location : Crystal::Location,
+      args = {} of String => Crystal::ASTNode
+
+    getter program, files
+
+    def initialize(@entrypoint : String,
+                   @def_locators : Array(String),
+                   @excludes : Array(String),
+                   @type_blocks : Bool,
+                   @type_splats : Bool,
+                   @type_double_splats : Bool,
+                   @prelude : String = "prelude",
+                   stats : Bool = false,
+                   progress : Bool = false,
+                   error_trace : Bool = false)
+      @entrypoint = File.expand_path(@entrypoint) unless @entrypoint.starts_with?("/")
+      @program = Crystal::Program.new
+      @files = Set(String).new
+      @warnings = [] of String
+
+      @program.progress_tracker.stats = stats
+      @program.progress_tracker.progress = progress
+      @program.show_error_trace = error_trace
+    end
+
+    # Run the entire typing flow, from semantic to file reformatting
+    def run : Hash(String, String)
+      semantic(@entrypoint, File.read(@entrypoint))
+
+      rets = {} of String => String
+
+      @warnings.each do |warning|
+        puts "WARNING: #{warning}"
+      end
+
+      @files.each do |file|
+        next unless File.file?(file)
+        source = File.read(file)
+        if typed_source = type_source(file, source)
+          rets[file] = typed_source
+        end
+      end
+
+      rets
+    end
+
+    # Take the entrypoint file (and its textual content) and run semantic on it.
+    # Semantic results are used to generate signatures for all defs that match
+    # at least one def_locator.
+    def semantic(entrypoint, entrypoint_content) : Nil
+      parser = program.new_parser(entrypoint_content)
+      parser.filename = entrypoint
+      parser.wants_doc = false
+      original_node = parser.parse
+
+      nodes = Crystal::Expressions.from([original_node])
+
+      if !@prelude.empty?
+        # Prepend the prelude to the parsed program
+        location = Crystal::Location.new(entrypoint, 1, 1)
+        nodes = Crystal::Expressions.new([Crystal::Require.new(@prelude).at(location), nodes] of Crystal::ASTNode)
+      end
+
+      program.normalize(nodes)
+
+      # And now infer types of everything
+      semantic_node = program.semantic nodes, cleanup: true
+
+      # We might run semantic later in an attempt to resolve defaults, don't display those stats or progress
```

Comment:
Per this comment, a method can be defined like:
```
def my_method(arg = "str")
```
In this case, semantic may run on the `"str"` (or whatever expression) in an attempt to figure out what it is. It's occasional that methods like these always get called with an argument that isn't the same type as the default, and then the compiler complains if the restriction ends up like:
```
def my_method(arg : Int32 = "str")
```